### PR TITLE
UN-191: Javascript enhancements for Insights

### PR DESCRIPTION
**Pull request by JamesWChan**
Hi @AshTNA,

This PR adds JavaScript enhancements for Insight pages. The enhancements are:

- Collapsable/expandable sections on mobile
- Progress indication within the sidebar navigation on desktop
- Smooth scrolling when using the navigation jumplinks on desktop

I've split the code into modules and placed the modules in an appropriate subfolder (e.g. all modules relating to the mobile accordion are in an accordion_functionality folder). I've tested the enhancements on desktop/mobile and in Chrome, Firefox, Safari, Edge, and IE11 (there are a few issues with IE11 but I'll fix them in a separate PR).

### CODE REVIEW

**Review by AshTNA — CHANGES_REQUESTED** [review]
Hi @JamesWChan 

Nice work. I commented on a few code things, and also discovered a few things:

- After resizing the page multiple times and then clicking a jumplink, i'm prevented from scrolling for a few seconds. I think extra event listeners might be being added everytime a resize happens.
- After resizing the screen, when clicking the jumplinks, it seems that the highlighting only reappears when the user scrolls again. I think this is related to the resize issue in the previous point.
- Is it OK that you can click multiple jumplinks quickly and make the page scroll all over the place? Maybe the jumplinks could be disabled while scrolling, or make it so a user scrolling manually themselves cancels any active/pending animation, so they can easily exit the auto scrolling? 
- The scroll animation, and the jumplink highlighting doesn't seem to work on Chrome on Windows at all for me. 
- The jumplinks are not stickied on IE11, is this OK?
- Will the insights pages be refactored? It looks like a lot of headings with no content become jumplinks, which is confusing on mobile. For example the "Introduction" heading on the Shakespeare page, or "The protest march" on Mangrove Nine. 
- This is a very minor thing, but on the mobile headings, when they expand should the triangle flip to show it's expanded? 

Thanks

File: scripts/src/insights.js [hunk 1]
```diff
@@ -3,11 +3,104 @@ import audio_tracking from "./modules/analytics/audio_tracking"
...
+import remove_aria_roles from "./modules/insights/accordion_functionality/remove_aria_roles";
+import remove_event from "./modules/insights/remove_event";
+import set_active from "./modules/insights/navigation/set_active";
+import throttle from "./modules/throttle";
+
 document.addEventListener('DOMContentLoaded', () => {
+    const $sectionHeadings = $(".section-separator__heading");
+    const $sectionContents = $(".section-content");
+    const $jumplinks = $(".jumplink");
+    const $windowWidth = $(window).width();
```
[the remark this answers is not in the document]
Comment (JamesWChan) [reply scripts/src/insights.js]: Decided just to remove the variable altogether and just went with `$(window).width()`

File: scripts/src/insights.js [hunk 2]
```diff
@@ -3,11 +3,104 @@ import audio_tracking from "./modules/analytics/audio_tracking"
...
+
+        // Detect if there are any expanded sections. If not, expand the first section.
+        open_first_section($sectionHeadings, $sectionContents);
+
+        // Add click and enter listeners for expanding/collapsing sections on small screens.
+        add_event($sectionHeadings, "click", function() {
+            accordion_functionality(this, $sectionHeadings, $sectionContents)
+        });
+        add_event($sectionHeadings, "keypress", function(e) {
+            if(e.which === 13 || e.which === 32) {
```
[the remark this answers is not in the document]
Comment (JamesWChan) [reply scripts/src/insights.js]: Good spot, thank you. Careless on my part. I've changed that now.

File: templates/insights/insights_index_page.html [hunk 3]
```diff
@@ -74,5 +74,10 @@ <h2>Discover all stories</h2>
 {% endblock %}
 
 {% block extra_js %}
+    <script 
+        src="https://code.jquery.com/jquery-3.6.0.min.js"
```
[the remark this answers is not in the document]
Comment (JamesWChan) [reply templates/insights/insights_index_page.html]: Done.

File: templates/insights/insights_page.html [hunk 4]
```diff
@@ -49,5 +49,10 @@ <h2 class="insights-container-related-content__heading">More insights</h2>
 {% endblock %}
 
 {% block extra_js %}
-     <script src="{% static 'scripts/insights.js' %}"></script>
+    <script 
+        src="https://code.jquery.com/jquery-3.6.0.min.js"
```
[the remark this answers is not in the document]
Comment (JamesWChan) [reply templates/insights/insights_page.html]: Done.

File: scripts/src/insights.js [hunk 5]
```diff
@@ -3,11 +3,104 @@ import audio_tracking from "./modules/analytics/audio_tracking"
...
+
+            // Detect if there are any expanded sections. If not, expand the first section.
+            open_first_section($sectionHeadings, $sectionContents);
+
+            // Add click and enter listeners for expanding/collapsing sections when the screen size is reduced.
+            add_event($sectionHeadings, "click", function() {
+                accordion_functionality(this, $sectionHeadings, $sectionContents)
+            });
+            add_event($sectionHeadings, "keypress", function(e) {
+                if(e.which === 13 || e.which === 32) {
```
[the remark this answers is not in the document]
Comment (JamesWChan) [reply scripts/src/insights.js:83]: Done.

**Comment by JamesWChan** [comment]
Hi @AshTNA,

Thanks for the review. Just to address some points:

- Regarding jumplinks in IE 11: IE11 doesn't support sticky positioning so I'd have to find a workaround. I did check with Gwyn and he said it was okay as it was but I can address this issue along with some IE11 bugs in another PR if it's required.
- Regarding headings with no content: editors will be adding content so empty headings won't appear. But we can always add a rule in Wagtail that will prevent a section from having a heading only (i.e. if a heading is added, content will need to be added too).

**Review by AshTNA — APPROVED** [review]

**Comment by AshTNA** [comment]
All good :+1:

**Pull request merged**